R/utils.R

```diff
@@ -24,3 +24,9 @@ silver    <- function(x) if (has_crayon()) crayon::silver(x)    else x
 bold      <- function(x) if (has_crayon()) crayon::bold(x)      else x
 italic    <- function(x) if (has_crayon()) crayon::italic(x)    else x
 underline <- function(x) if (has_crayon()) crayon::underline(x) else x
+
+lifecycle_abort <- function(x, env = parent.frame()) {
+  x <- glue::trim(x)
+  x <- paste0("[lifecycle] ", x)
+  abort(glue::glue(x, .envir = env), class = "lifecycle_error")
+}
```

Comment:
* I don't have strong feelings about that; I just added it because it seemed like an internal lifecycle error was a meaningful concept
* I think quite a few packages have `abort()` wrappers that call glue.
* I don't like "internal" because it's not clear what it's internal to — I don't particularly care about what it looks like except I think it should include "lifecycle".

Reply:
I have somewhat strong feelings about avoid classes for uninteresting errors because with classes they become part of the API to some extent.

We could pick up the current namespace from the caller env:

```r
stop_internal <- function(fn, msg) {
  ns <- topenv(caller_env())
  if (is_namespace(ns)) {
    what <- paste0(ns_env_name(ns), " error")
  } else {
    what <- "error"
  }

  abort(sprintf("Internal %s in `%s()`: %s", what, fn, msg))
}
```

Reply:
That seems reasonable for a general rlang function. I still think it's worth baking `glue()` in — in fact, I wish it was baked into `abort()`, if we could figure out how to resolve the dependency issue.

Reply:
We conditionally depend on glue for the LHS of `:=` in dots capture so we could support it in `abort()` as well. Then if a package wishes to use glue in `abort()`, it would need to make sure glue is available by adding it to `Imports`. I opened an issue for this at https://github.com/r-lib/rlang/issues/1090.

Reply:
Ok, take another look now.